src/typechecker.erl

```diff
@@ -2783,7 +2783,11 @@ type_check_comprehension_in(Env, ResTy, bc, Expr, P, []) ->
                             [{integer, erl_anno:new(0), 0},
                              {integer, erl_anno:new(0), 1}]};
                  _ ->
-                     throw({type_error, bc, P, ResTy})
+                     case subtype({type, erl_anno:new(0), binary, []}, ResTy, Env) of
```

Comment:
This covers ResTy `iodata()`, but what if ResTy is a union containing `bitstring()` or another bit type such as `<<_:7, _:_*3>>`? I think the general solution would be do like in the list comprehension case and create a function `expect_bitstring_type`.

Another solution: I think we can get rid of all the `expect_` functions though if we split any union (e.g. `iodata()` after normalizing it) before this function is called. I'm not sure if there is a performance concern for doing this, @josefs? Personally I'm more concerned about the code complexity than the performance for these things.

Reply:
Ok, thinking about that, I came up with this other testcase, which fails for my current solution:
```erlang
-spec bitstring_fun() -> <<_:7, _:_*3>> | string().
bitstring_fun() -> << <<1:N>> || N <- lists:seq(3, 12)>>.
```

I think I'm coming for a solution for that, with the idea of splitting the unions, but the code is getting a bit uglier at least from my perspective. I'll push it here when I have it, but maybe I could put it on a different PR so that you can choose the best solution, maybe?

Reply:
An even crazier example
```erlang
-type nested() :: string() | bitstring() | binary().
-spec nested_bitstrings() -> nested() | boolean() | pid().
nested_bitstrings() ->
    << <<1:N>> || N <- lists:seq(3, 12)>>.
```
:smiley: